internal/provider/data_source.go

```diff
@@ -31,105 +48,111 @@ func dataSource() *schema.Resource {
+					listvalidator.SizeAtLeast(1),
 				},
-				MinItems: 1,
 			},
 
-			"working_dir": {
+			"working_dir": schema.StringAttribute{
 				Description: "Working directory of the program. If not supplied, the program will run " +
 					"in the current directory.",
-				Type:     schema.TypeString,
 				Optional: true,
-				Default:  "",
 			},
 
-			"query": {
+			"query": schema.MapAttribute{
 				Description: "A map of string values to pass to the external program as the query " +
 					"arguments. If not supplied, the program will receive an empty object as its input.",
-				Type:     schema.TypeMap,
-				Optional: true,
-				Elem: &schema.Schema{
-					Type: schema.TypeString,
-				},
+				ElementType: types.StringType,
+				Optional:    true,
 			},
 
-			"result": {
+			"result": schema.MapAttribute{
 				Description: "A map of string values returned from the external program.",
-				Type:        schema.TypeMap,
+				ElementType: types.StringType,
+				Computed:    true,
+			},
+
+			"id": schema.StringAttribute{
+				Description: "The id of the data source. This will always be set to `-`",
 				Computed:    true,
-				Elem: &schema.Schema{
-					Type: schema.TypeString,
-				},
 			},
 		},
 	}
 }
 
-func dataSourceRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+func (n *externalDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var config externalDataSourceModelV0
 
-	programI := d.Get("program").([]interface{})
-	workingDir := d.Get("working_dir").(string)
-	query := d.Get("query").(map[string]interface{})
+	diags := req.Config.Get(ctx, &config)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	var program []string
 
-	program := make([]string, 0, len(programI))
+	diags = config.Program.ElementsAs(ctx, &program, false)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
-	for _, programArgRaw := range programI {
-		programArg, ok := programArgRaw.(string)
+	filteredProgram := make([]string, 0, len(program))
 
-		if !ok || programArg == "" {
+	for _, programArgRaw := range program {
+		if programArgRaw == "" {
 			continue
 		}
 
-		program = append(program, programArg)
+		filteredProgram = append(program, programArgRaw)
 	}
 
-	if len(program) == 0 {
-		return diag.Diagnostics{
-			{
-				Severity:      diag.Error,
-				Summary:       "External Program Missing",
-				Detail:        "The data source was configured without a program to execute. Verify the configuration contains at least one non-empty value.",
-				AttributePath: cty.GetAttrPath("program"),
-			},
-		}
+	if len(filteredProgram) == 0 {
+		resp.Diagnostics.AddAttributeError(
+			path.Root("program"),
+			"External Program Missing",
+			"The data source was configured without a program to execute. Verify the configuration contains at least one non-empty value.",
+		)
+		return
+	}
+
+	var query map[string]string
```

Comment:
Similarly with `query`, e.g.

```terraform
data "external" "test" {
  query = {
    test = null
  }
}
```

And using `map[string]types.String` 👍 

We could potentially consider adding explicit validation in the schema to prevent null elements in a future major version or explicitly stating they will always be converted to an empty string.